### PR TITLE
Analyze Target object field types to avoid caching refs to long-living objects

### PR DESCRIPTION
**Pull request by SergeiPavlov**
+ refactoring

### CODE REVIEW

File: Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs [hunk 1]
```diff
@@ -118,80 +121,90 @@ public DelayedScalarQuery<TResult> CreateDelayedQuery<TResult>(Func<QueryEndpoin
...
-      PutCachedQuery(parameterizedQuery);
+      if (cacheable) {
+        PutCachedQuery(parameterizedQuery);
+      }
       return parameterizedQuery;
     }
 
-    private void AllocateParameterAndReplacer()
+    // Returns true is query is cacheable (target contains only simple-type captured vars).
+    private bool AllocateParameterAndReplacer()
```
Comment (botinko) [inline Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs:158]: @commit-n-run Could you review this approach?

File: Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs [hunk 2]
```diff
@@ -118,80 +121,90 @@ public DelayedScalarQuery<TResult> CreateDelayedQuery<TResult>(Func<QueryEndpoin
...
+
+        var expressionType = expression.Type;
+        if (expressionType.IsClosure()) {
+          return expressionType == closureType
+            ? Expression.MakeMemberAccess(Expression.Constant(queryParameter, parameterType), valueMemberInfo)
+            : throw new NotSupportedException(string.Format(Strings.ExExpressionDefinedOutsideOfCachingQueryClosure, expression));
         }
-        return null;
+
+        return expressionType.IsAssignableFrom(closureType)
```
Comment (commit-n-run) [inline Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs:182]: This heavy rewrite of expressions makes it hard to understand what were the actual changes

Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs:182]: Replaced by more simple https://github.com/DataObjects-NET/dataobjects-net/pull/225
Without refactoring

**Comment by SergeiPavlov** [comment]
Replaced by https://github.com/DataObjects-NET/dataobjects-net/pull/223

**Pull request closed without merging**